New command pep517.meta

jaraco opened the pull request:
As outlined in #26, this command implements a ~`pep517.build-meta`~`pep517.meta` command, allowing for building metadata for a source directory.

The new ~`build_meta`~`meta` module also exposes an API for retrieving the metadata as an in-memory zip file allowing consumers to receive metadata in-process without thread safety and other race concerns.

Additionally, this request adds a `compat_system` function which, given a source directory, produces a "build system", either by loading it from pyproject.toml or falling back to "setuptools (legacy)".

This proposed technique fulfills the needs of `importlib_metadata` or other consumers that wish to load metadata from PEP 517 projects on disk, or if `compat_system()` is passed for `system`, for most any project.

Edit: This technique additionally introduces a `load()` function for loading an `importlib_metadata.Distribution` object from a source directory.

One downside I'd like to still investigate:

- [x] When generating the metadata, a lot of output is emitted. It would be nice to be able to suppress that output.

jaraco commented:
The code in b35a8dd is currently released to https://m.devpi.net/jaraco/pep517-meta.

jaraco commented:
See [importlib_metadata!67](https://gitlab.com/python-devs/importlib_metadata/merge_requests/67) for the corresponding importlib_metadata work.

pfmoore commented:
If I understand correctly, this implements a proposed extension to PEP 517? If that's true, then I think that we should hold off on merging this PR until that proposed extension is accepted,

jaraco commented:
> this implements a proposed extension to PEP 517?

No. This PR only implements APIs around the already specified hooks defined in PEP 517, namely [prepare_metadata_for_build_wheel](https://www.python.org/dev/peps/pep-0517/#prepare-metadata-for-build-wheel).

This new command and API follow the patterns found in `pep517.build` and `pep517.check` to present comparable functionality for metadata.

It does go one step further. Based on the discussion and recommendation in pypa/packaging-problems#226, the `build_meta_as_zip` function builds the metadata in a temporary directory, zips it up in-memory, and then returns that in-memory zipfile.

But as you can see, each of these functions are disentangled from the more fundamental functionality (building the metadata in some target directory).

pfmoore commented:
>> this implements a proposed extension to PEP 517?
>
> No.

OK, cool. My apologies, I got the idea it was an extension from [this comment](https://github.com/pypa/pep517/issues/26#issuecomment-446723506). I must have misread the thread (I was skimming quickly, I don't have time right now for a full review).

gaborbernat commented on pep517/build-meta.py:
```diff
@@ -0,0 +1,3 @@
+from . import build_meta
+
+__name__ == '__main__' and build_meta.main()
```
wow I've never seen this cryptic way of doing this 🤔

jaraco replied:
I agree it's idiosyncratic, but I've run it by other seasoned devs and gotten at least a nod of endorsement about the one-line approach, so I've started using it. I personally prefer it to the traditional approach for two reasons: (1) it fits on one line, and (2) it encourages exactly one expression for `__main__`. What I'd really prefer is a builtin that does what [autocommand](https://pypi.org/project/autocommand) does - decorate a function to be run as main. I use that wherever I can rely on Python 3 and third-party dependencies.

pfmoore replied:
I'm -1 on this. Let's keep to the standard `if __name__ == '__main__'` idiom. Why make it harder for future maintainers to understand for little benefit?

gaborbernat replied:
I definitely find it hard to read, but that's only me perhaps.

jaraco replied:
I updated this and ultimately removed the file entirely after I found that `pep517.meta` was a more suitable name for both the command and the API module.

jaraco commented:
Here's an example of this functionality in action:

```
~ $ pip-run -i https://m.devpi.net/jaraco/pep517-meta pep517                                                                                                                   
Looking in indexes: https://m.devpi.net/jaraco/pep517-meta
Collecting pep517
  Downloading https://m.devpi.net/jaraco/pep517-meta/%2Bf/1f7/c95a2a4d63948/pep517-0.5.0-py2.py3-none-any.whl (49kB)
     |████████████████████████████████| 51kB 440kB/s 
Collecting pytoml (from pep517)
Installing collected packages: pytoml, pep517
Successfully installed pep517-0.5.0 pytoml-0.1.20
>>> Python 3.7.3 (v3.7.3:ef4ec6ed12, Mar 25 2019, 16:52:21) 
[Clang 6.0 (clang-600.0.57)] on darwin
Type "help", "copyright", "credits" or "license" for more information.

>>> import pep517.build_meta
>>> import functools
>>> import os
>>> builder = functools.partial(pep517.build_meta.build_meta, source_dir=os.path.expanduser('~/m/jaraco.functools'))
>>> meta_zip = pep517.build_meta.build_meta_as_zip(builder)
Collecting setuptools>=34.4
  Using cached https://files.pythonhosted.org/packages/ec/51/f45cea425fd5cb0b0380f5b0f048ebc1da5b417e48d304838c02d6288a1e/setuptools-41.0.1-py2.py3-none-any.whl
Collecting wheel
  Using cached https://files.pythonhosted.org/packages/bb/10/44230dd6bf3563b8f227dbf344c908d412ad2ff48066476672f3a72e174e/wheel-0.33.4-py2.py3-none-any.whl
Collecting setuptools_scm>=1.15
  Using cached https://files.pythonhosted.org/packages/1d/70/97966deebaeeda0b81d3cd63ba9f8ec929b838871ed17476de9d8159db3e/setuptools_scm-3.3.3-py2.py3-none-any.whl
Installing collected packages: setuptools, wheel, setuptools-scm
Successfully installed setuptools setuptools-scm wheel
Collecting wheel
  Using cached https://files.pythonhosted.org/packages/bb/10/44230dd6bf3563b8f227dbf344c908d412ad2ff48066476672f3a72e174e/wheel-0.33.4-py2.py3-none-any.whl
Collecting setuptools_scm>=1.15.0
  Using cached https://files.pythonhosted.org/packages/1d/70/97966deebaeeda0b81d3cd63ba9f8ec929b838871ed17476de9d8159db3e/setuptools_scm-3.3.3-py2.py3-none-any.whl
Installing collected packages: wheel, setuptools-scm
Successfully installed setuptools-scm-3.3.3 wheel-0.33.4
running dist_info
creating /var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/tmp04lgx4m_/jaraco.functools.egg-info
writing /var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/tmp04lgx4m_/jaraco.functools.egg-info/PKG-INFO
writing dependency_links to /var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/tmp04lgx4m_/jaraco.functools.egg-info/dependency_links.txt
writing requirements to /var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/tmp04lgx4m_/jaraco.functools.egg-info/requires.txt
writing top-level names to /var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/tmp04lgx4m_/jaraco.functools.egg-info/top_level.txt
writing manifest file '/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/tmp04lgx4m_/jaraco.functools.egg-info/SOURCES.txt'
writing manifest file '/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/tmp04lgx4m_/jaraco.functools.egg-info/SOURCES.txt'
creating '/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/tmp04lgx4m_/jaraco.functools.dist-info'
>>> meta_zip
<zipfile.ZipFile file=<_io.BytesIO object at 0x10f9985c8> mode='w'>
```

Another thing I'm considering is instead of stopping at generating the zip file, perhaps `pep517` could own [this functionality](https://gitlab.com/python-devs/importlib_metadata/blob/35189f591fa736da4a28cdb20b34cf9be02c2918/importlib_metadata/__init__.py#L200-206). Then, `importlib_metadata` would facilitate the metadata parsing, but `pep517` could own the metadata loading logic. That feels like a better separation of concerns, especially as `importlib_metadata` enters the stdlib and can't reliably depend on `pep517`.

gaborbernat commented on pep517/build_meta.py:
```diff
@@ -0,0 +1,82 @@
+    return zip_file
+
+
+def build_meta_as_zip(builder=build_meta):
```
this is nowhere used, what's the goal with this?

jaraco replied:
The latest commits use this code. It's meant to supply the metadata as the return value of the function. Because the metadata is inherently a multi-file structure, it has to be a zip file.

jaraco commented:
Also, as is demonstrated by failures I discovered above, we should add tests.

- [ ] Add tests for this functionality.

gaborbernat commented on pep517/build_meta.py:
```diff
@@ -0,0 +1,82 @@
+        return dir_to_zipfile(out_dir)
+
+
+parser = argparse.ArgumentParser()
```
would prefer this to be not global (aka let's put it inside a function, and call it at the start of main); will make it easier to test

jaraco replied:
I agree, except this code follows the pattern of the `build` and `check` modules. It's more important for these to be consistent. If we wish to change the pattern (a reasonable request), I suggest doing that subsequently to accepting this addition.

gaborbernat replied:
You're correct.

gaborbernat commented on pep517/dirtools.py:
```diff
@@ -0,0 +1,25 @@
+
+
+@contextlib.contextmanager
+def tempdir():
```
this deserves some documentation

jaraco replied:
Done.

gaborbernat commented on pyproject.toml:
```diff
@@ -10,6 +10,8 @@ home-page = "https://github.com/takluyver/pep517"
 description-file = "README.rst"
 requires = [
     "pytoml",
+    "importlib_metadata",
```
why is this a build dependency? 🤔

jaraco replied:
I thought that was an install dependency.

pfmoore replied:
This one's the flit metadata section, not the build-system section.

It's a bit unfortunate the name's used in 2 places for 2 different things :-(

jaraco commented:
As you can see, tests are failing on Python 2.7 due to `flit` not being available on Python 2.7 (and as a result, it's not possible to build the metadata for `pep517` on Python 2.7). I guess that's an xfail?

jaraco commented:
In e3138f47e3a921abe7627087c1a57ab82023c79b and d27a2eceb41f403e437b06e314e2e81ed6dbe7f0, I added a test and updated the implementation to ensure that the pip install and build steps don't emit any output. The latter change has implications for other parts of the library that also call `pip_install` (their output will be captured/suppressed also). For this reason alone, I'm uneasy about this change.

I'd like feedback from others monitoring this thread on those concerns.

jaraco commented:
I feel a lot better about this latest implementation. It suppresses the output from the `pip_install` invocation by redirecting the output to the log.

gaborbernat commented on pep517/build.py:
```diff
@@ -3,25 +3,50 @@
+    """
+    Ensure build system has the requisite fields.
+    """
+    required = {'requires', 'build-backend'}
```
seems overzealous, can't we fallback for the build-backend key to setuptools legacy as pip does?

gaborbernat replied:
I would say even for requires we should just inject ``setuptools >= 41.0.0`` and ``wheel >= 0.30.0``.

jaraco replied:
This function is simply about ensuring the requisite fields are there. There is another function, `compat_system`, that supplies the fallback. If we could say that "there is no case where the fallback would not be desirable", then I agree, the fallback should be derived unconditionally and there's no need to validate. But since the existing behavior is to fail when an invalid backend is supplied, I wanted to retain that behavior for existing uses and for the use-case herein, support the fallback.

jaraco replied:
I've filed #49 to follow up on this consideration separately.

gaborbernat commented on pep517/build.py:
```diff
@@ -3,25 +3,50 @@
+    """
+    Load the build system from a source dir (pyproject.toml).
+    """
+    pyproject = os.path.join(source_dir, 'pyproject.toml')
```
again let's fallback to old things rather than fail if missing 👍

jaraco replied:
And the same comment applies here. The `compat_system` provides this functionality.

gaborbernat commented on pep517/envbuild.py:
```diff
@@ -90,9 +90,14 @@ def pip_install(self, reqs):
             sys.executable, '-m', 'pip', 'install', '--ignore-installed',
-            '--prefix', self.path] + list(reqs))
+            '--prefix', self.path] + list(reqs)
+        check_call(
```
will this print out the output if the call fails? 🤔

jaraco replied:
It will not. Previously, the output would be emitted unconditionally. Now the output will be redirected to the log, where it can be controlled based on logging settings. I think this library probably needs to step back and specify what callers should expect about output across various operations.

jaraco replied:
I'm hoping not to have to solve that problem here but only provide a reasonable behavior and avoid emitting the output for API callers.

jaraco replied:
This issue is being addressed more generally in #13.

gaborbernat commented on tests/test_meta.py:
```diff
@@ -0,0 +1,38 @@
+from pep517 import meta
+
+
+pep517_needs_python_3 = pytest.mark.xfail(
```
what 🤔 why does this requires python 3? 🤔

jaraco replied:
`pep517` is built with [flit](https://pypi.org/project/flit), which is a Python 3 only package. Because these tests are relying on this project's own repo/metadata to run the tests, those tests therefore have the same constraint. It's the same reason you can't install pep517 from source on Python 2.7:

```
$ python2.7 -m pip-run --no-binary=:all: --no-deps pep517                                                                                                                    
DEPRECATION: Python 2.7 will reach the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 won't be maintained after that date. A future version of pip will drop support for Python 2.7.
Collecting pep517
  Downloading https://files.pythonhosted.org/packages/79/27/1e6a8ba83dab2b243e1020e340552c129b0076fbb5518e0c2a9c44c7d056/pep517-0.5.0.tar.gz
  Installing build dependencies ... error
  Complete output from command /usr/local/opt/python@2/bin/python2.7 /usr/local/lib/python2.7/site-packages/pip install --ignore-installed --no-user --prefix /private/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/pip-build-env-jCiSaR/overlay --no-warn-script-location --no-binary :all: --only-binary :none: -i https://pypi.org/simple -- flit:
  DEPRECATION: Python 2.7 will reach the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 won't be maintained after that date. A future version of pip will drop support for Python 2.7.
  Collecting flit
    Could not find a version that satisfies the requirement flit (from versions: )
  No matching distribution found for flit
  
  ----------------------------------------
Command "/usr/local/opt/python@2/bin/python2.7 /usr/local/lib/python2.7/site-packages/pip install --ignore-installed --no-user --prefix /private/var/folders/c6/v7hnmq453xb6p2dbz1gqc6rr0000gn/T/pip-build-env-jCiSaR/overlay --no-warn-script-location --no-binary :all: --only-binary :none: -i https://pypi.org/simple -- flit" failed with error code 1 in None
```

These tests could be re-written to use mocks instead of the local pep517 checkout, but I'm satisfied with this approach.

gaborbernat replied:
Should not be that hard to run the tests on a setuptools simple project, not 🤔 flit does generate wheel/sdist that can be installed on Python 2 🤔

jaraco replied:
I do think it would be worthwhile for coverage reasons to retrieve metadata from a non PEP 517 source package.

> flit does generate sdist that can be installed on Python 2

Can you demonstrate this behavior? Because the pep517 sdist seems not to have that feature.

gaborbernat replied:
Eh, looking through the CI and tests there's no such tests or coverage report yet. As such for now probably not worth it. Let's release and a subsequent PR can improve the CI to generate coverage, and then we can worry about this.

gaborbernat commented:
@pfmoore can we land this?

pfmoore commented:
I'm going to leave that decision to @takluyver as owner of this project. My only real interest is in the wrappers, which pip uses.

takluyver commented on pep517/build.py:
```diff
@@ -3,25 +3,50 @@
+    Ensure build system has the requisite fields.
+    """
+    required = {'requires', 'build-backend'}
+    if required > set(system):
```
I don't think this does quite what you expect. Try

```python
if not (required <= set(system)):
```

`>` catches if the keys are a strict subset of the required ones. So it works if you miss one out completely, but if you miss one out and add another - e.g. by making a typo - the keys are no longer a subset.

`<=` checks that the keys are a non-strict superset of the required ones, which is what we care about.

jaraco replied:
Good catch. I was able to replicate your findings with this patch:

```
diff --git a/conftest.py b/conftest.py
new file mode 100644
index 0000000..4f1eca8
--- /dev/null
+++ b/conftest.py
@@ -0,0 +1,7 @@
+import sys
+
+collect_ignore = ['pep517/_in_process.py']
+
+
+if sys.version_info < (3,):
+    collect_ignore.append('pep517/build.py')
diff --git a/pep517/build.py b/pep517/build.py
index 781f450..8aafc1c 100644
--- a/pep517/build.py
+++ b/pep517/build.py
@@ -17,6 +17,31 @@ log = logging.getLogger(__name__)
 def validate_system(system):
     """
     Ensure build system has the requisite fields.
+
+    >>> validate_system({'build-backend': None, 'requires': None})
+    >>> validate_system({'build-backend': None, 'requires': None,
+    ...     'other': None})
+
+    >>> validate_system(dict(requires=None))
+    Traceback (most recent call last):
+    ...
+    ValueError: Missing required fields: {'build-backend'}
+
+    >>> validate_system({'build-backend': None})
+    Traceback (most recent call last):
+    ...
+    ValueError: Missing required fields: {'requires'}
+
+    >>> validate_system({})
+    Traceback (most recent call last):
+    ...
+    ValueError: Missing required fields: {'...', '...'}
+
+    >>> validate_system({'build-backend': None, 'other': None})
+    Traceback (most recent call last):
+    ...
+    ValueError: Missing required fields: {'requires'}
+
     """
     required = {'requires', 'build-backend'}
     if required > set(system):
diff --git a/pytest.ini b/pytest.ini
index 5c37966..b5835b8 100644
--- a/pytest.ini
+++ b/pytest.ini
@@ -1,2 +1,2 @@
 [pytest]
-addopts=--flake8
+addopts=--flake8 --doctest-modules
```

(only runs on Python 3)

However, there are too many complications with using doctests on this project, so I'm abandoning that work.

jaraco replied:
Fixed in 1f90e57.

jalanb replied:
@jaraco 

Just passing through, following a link from "GitHub Explore the week", and having decided to concentrate more on doctests (at work): may I ask what complications arose here with using them?

takluyver commented on pep517/build.py:
```diff
@@ -3,25 +3,50 @@
+    required = {'requires', 'build-backend'}
+    if required > set(system):
+        missing = required - set(system)
+        message = "Missing required fields: {missing}".format(**locals())
```
I'd rather lose the `locals()` introspection trick. It's a distraction if you're not used to reading it, and it doesn't save any characters.

jaraco replied:
I prefer this technique because it most closely translates to the syntax for f-strings, which I'd use here once available. I see the `**locals()` trick as a wart that's a valuable reminder to use f-strings. But since you feel strongly enough, I can omit it.

takluyver replied:
Thanks; I understand the point, but I think the downside of the wart outweighs the reminder to use f-strings, which aren't a big win here anyway.

jaraco replied:
Fixed in bdad30a.

takluyver commented on pep517/meta.py:
```diff
@@ -0,0 +1,90 @@
+
+def load(root):
+    """
+    Load the metadata from root.
```
This could be more descriptive. What is `root`? (I can see from the code that it's a source directory, but it's not obvious from the docstring)

The name also doesn't seem very meaningful - it doesn't load metadata, it builds metadata and then returns a wrapper around it to conveniently access it.

jaraco replied:
Does 7d568e5 clarify?

takluyver commented on tests/test_meta.py:
```diff
@@ -0,0 +1,38 @@
+
+pep517_needs_python_3 = pytest.mark.xfail(
+    'sys.version_info < (3,)',
+    reason="pep517 cannot be built on Python 2",
```
I'm working towards resolving this, but only in odd scraps of time at the moment.

jaraco commented:
Thanks for the review. All good points. I'll follow up later.

jaraco commented on pep517/build.py:
```diff
@@ -3,25 +3,50 @@
+        system = load_system(source_dir)
+    except Exception:
+        system = {}
+    system.setdefault('build-backend', 'setuptools.build_meta')
```
I can see your point. My instinct was, since pep517 never had the legacy behavior, it's not necessary to drag it into here. That is, the "meta" command will never support that legacy assumption. That means it won't support those projects reliant on that behavior. If I recall, when `__legacy__` was created, it was meant for pip only (and that's why it's `__`-prefixed). I'd prefer to start with this constrained approach and relax the constraint later if demand and consideration indicate.

takluyver replied:
I suspect we'll soon end up having to use the 'legacy' backend, because the people building a project through this mechanism are not necessarily the ones deciding what assumptions that project makes.

@pfmoore @gaborbernat (and anyone else): any preference on this?

gaborbernat replied:
I'd go in +1 to make it setuptools.build_meta:__legacy__ the default.

jaraco replied:
Updated in 205d128.

jaraco commented on pep517/build.py:
```diff
@@ -3,25 +3,50 @@
-    finally:
-        shutil.rmtree(td)
+        system = load_system(source_dir)
+    except Exception:
```
I do intend to trap both OSError and KeyError here... because the presence of a pyproject.toml without a `build-system` is comparable to no pyproject.toml (for this concern). I'd be tempted to say a malformed pyproject.toml should be treated as having no declared build system, though I do sympathize with your instinct here.

The problem I have with specifying `(OSError, KeyError)` in the `compat_system` function is it couples the internal implementation details of `load_system` with `compat_system`. I guess that's okay as one is meant to wrap the other, but it does make me a little uneasy.

Addressed in 0da9696.

jaraco commented:
I believe I've addressed all of the concerns that @takluyver raised. Any remaining concerns or comments?

takluyver commented:
Thanks!

gaborbernat commented:
@takluyver want to do a release with this? thanks!

takluyver commented:
There you go, 0.6.0 now out.

takluyver commented:
PS @jaraco or anyone: if you got a notification about the deploy failure, don't worry. I had forgotten that the releases were automated, and my personal release habit is to upload to PyPI first and then create the tag, so it failed because the packages were already on PyPI.
